terraform/diff.go

```diff
@@ -373,6 +373,11 @@ type InstanceDiff struct {
 	// mean to be used for additional data a resource may want to pass through.
 	// The value here must only contain Go primitives and collections.
 	Meta map[string]interface{}
+
+	// Stub should be set when this diff exists only for purposes of providing a
+	// diff to various pre-plan or dry-run steps in the graph. A diff with this
+	// enabled should not be acted on in the plan.
+	Stub bool
```

Comment:
I think what I was going for here was just to make sure we were addressing the specific problem of resources being counted twice without impacting anything else - so I wanted to make sure that this was as granular as possible.

Looks like there are a few hooks right now:
 * [The count hook](https://github.com/hashicorp/terraform/blob/master/backend/local/hook_count.go) - The problematic hook
 * [The stop hook](https://github.com/hashicorp/terraform/blob/master/terraform/hook_stop.go) - Looks like every hook goes to a hook that halts execution if stopped
 * [The debug hook](https://github.com/hashicorp/terraform/blob/master/terraform/debug.go#L262) - Looks like it's dumping the diff to a debug file - which looks like it might be unused at this time - see [here](https://github.com/hashicorp/terraform/blob/5d19148f4751f137c8eb8797649c0963e6e854ff/command/plan.go#L116) and [here](https://github.com/hashicorp/terraform/blob/5d19148f4751f137c8eb8797649c0963e6e854ff/command/apply.go#L128) - unless its functionality is being used in a different capacity
 * [The mock hook](https://github.com/hashicorp/terraform/blob/master/terraform/hook_mock.go) - mainly just to track what hooks are called during testing, from just looking at how the hook is structured

As a quick assessment, the only hook that I could see being impacted by skipping the post-diff hook altogether would be the stop hook (might be a stretch though). My concern with skipping the hook altogether is that by simply structuring it as a modular thing, we're implying that hooks carry out a number of behaviours that are transparent to the something like `EvalDiff` altogether, so the eval should not make any assumptions on whether or not its functionality is useful in a specific scenario like if the diff is not a diff that should be acted upon.

I agree that it seems weird, it does to me too, and I for sure feel a little dirty overloading `InstanceDiff` like that, but in light of the above it seemed like the less of the two evils. Maybe I'm overthinking it. :wink: 

If you think it's safe though then definitely we could do it. I kind of wonder though, by that logic, what about `PreDiff`? Should that be skipped too? Just looked in the code too and I don't think that hook is doing *anything* meaningful at all.